pkg/commands/credentials.go

```diff
@@ -0,0 +1,299 @@
+		"LAZYGIT_LISTENER="+currentListener,
+
+		"GIT_ASKPASS="+ex, // tell git where lazygit is located so it can ask lazygit for credentials
+
+		"LANG=en_US.UTF-8",   // Force using EN as language
+		"LC_ALL=en_US.UTF-8", // Force using EN as language
+	)
+
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		outString := string(out)
+		if len(outString) == 0 {
+			end <- err
+			return
+		}
+		end <- errors.New(outString)
+		return
+	}
+	end <- nil
+}
+
+// runServer starts the server that waits for events from the lazygit client
+func runServer(serverStartedChan chan struct{}, end chan error, hostPort, currentListener string) {
+	serverRunning := false
+
+	addy, err := net.ResolveTCPAddr("tcp", "127.0.0.1:"+hostPort)
+	if err != nil {
+		end <- err
+		return
+	}
+
+	inbound, err := net.ListenTCP("tcp", addy)
+	if err != nil {
+		end <- err
+		return
+	}
+
+	go func() {
+		<-end
+		if serverRunning {
+			inbound.Close()
+		}
+	}()
+
+	listener := new(Listener)
+
+	// every listener needs a different name it this is not dune rpc.RegisterName will error
+	err = rpc.RegisterName("Listener"+currentListener, listener)
+	if err != nil {
+		end <- err
+		return
+	}
+
+	serverStartedChan <- struct{}{}
+
+	serverRunning = true
+	rpc.Accept(inbound)
+	serverRunning = false
+}
+
+// GetFreePort returns a free port that can be used by lazygit
+func GetFreePort() string {
+	checkFrom := 5000
+	toReturn := ""
+	for {
+		checkFrom++
+		check := fmt.Sprintf("%v", checkFrom)
+		if IsFreePort(check) {
+			toReturn = check
+			break
+		}
+	}
+	return toReturn
+}
+
+// IsFreePort return true if the port if not in use
+func IsFreePort(port string) bool {
+	conn, err := net.Dial("tcp", "127.0.0.1:"+port)
+	if err == nil {
+		go conn.Close()
+		return false
+	}
+	return true
+}
+
+// SetupClient sets up the client
+// This will be called if lazygit is called through git
+func SetupClient() {
+	port := os.Getenv("LAZYGIT_HOST_PORT")
+	listener := os.Getenv("LAZYGIT_LISTENER")
+
+	privateKey, publicKey, err := generateKeyPair()
+	if err != nil {
+		return
+	}
+
+	client, err := rpc.Dial("tcp", "127.0.0.1:"+port)
+	if err != nil {
+		return
+	}
+
+	var data *EncryptedMessage
+	err = client.Call("Listener"+listener+".Input", InputQuestion{
+		Question:  os.Args[len(os.Args)-1],
+		Listener:  listener,
+		PublicKey: publicKey,
+	}, &data)
+	client.Close()
+	if err != nil {
+		return
+	}
+
+	msg, err := decryptMessage(privateKey, *data)
+	if err != nil {
+		return
+	}
+
+	fmt.Println(msg)
+}
+
+// HasLGAsSubProcess returns true if lazygit is a child of this process
+func HasLGAsSubProcess() bool {
+	if !ps.Supported() {
+		return true
+	}
+
+	lgHostPid := os.Getpid()
+	list, err := ps.Processes()
+	if err != nil {
+		return false
+	}
+procListLoop:
+	for _, proc := range list {
+		procName := proc.Executable()
+		if procName != "lazygit" && procName != "lazygit.exe" {
+			continue
+		}
+		parrent := proc.PPid()
+		for {
+			if parrent < 30 {
```

Comment:
why are we checking for the PID being less than 30 here?

Reply:
go-ps has a bug where it panics if it tries to read the ppid of a process that has no ppid. the `.ppid` method has a return error but seems to be useless :).  

On linux every process has eventually a sub process that is either 1 or 2, but... on windows this is a pure mess. Windows has programs with a PID over the 10000 that panic when i try to get there ppid.  
Though when i ask for the sub process of lazygit it usually* will go back to a PPID under the 30 if looked over every sub process.  
I said usually because everything can be different every time so also it might loop over an item that has a PID over the 30 but is a lowest program.   After some testing 30 seems to be a nice sweet spot where go-ps doesn't panic (at least for me)